### ISSUE TITLE
enhance the pipelineloop controller for embedded custom task spec

### ISSUE THREAD

**Issue by Tomcli** [kubeflow/kfp-tekton#605]
> Thanks for the information, from the diagram user will get the response once the pipelinerun created successfully no matter the Run CR creation status? I think we need enhance it later.

@pugangxa Yes right now the Run CR just created without any validation. From the [TEP 61](https://github.com/tektoncd/community/pull/393), the recommended feedback is the have some kind of async validation from the custom controller. So the not all the RUN CRs will be pending when one of the custom task validation webhooks is failing. Let me open another issue to track these enhancement.

To enhance the pipelineloop with embedded spec, we need
1. async validation from the custom controller
2. have custom task timeout when the custom controller is not available. https://github.com/tektoncd/pipeline/pull/3976

_Originally posted by @Tomcli in https://github.com/kubeflow/kfp-tekton/issues/602#issuecomment-852299583_

**Comment by Tomcli** [comment]
The only tricky thing on this enhancement is that with embedded spec, the custom controller only can see the task spec when the RUN CR is submitted/created. Whereas with taskRef, the spec are created before the pipelineRun is running which helps catching syntax/spec errors early on.

**Comment by Tomcli** [comment]
If there is a HTTP validation API or binary for the whole pipelinerun on custom task, the KFP server can use that API/binary to verify the spec before creating the pipelineRun CR. However, since every new custom task api/kind is host on a different custom task server, this approach is not viable unless we have all custom task APIs/Kinds being implemented in one single controller. In which is sort of against the TEP/experimental process if we want our custom task be merge into Tekton.

**Comment by pugangxa** [comment]
I prefer to not do it in one controller. At the worst case we did nothing, I think seeing the task spec when the RUN CR submitted/created is still acceptable, as long as it can fail and let user know what happened. It's the same process as the normal tasks.

**Comment by ScrapCodes** [comment]
Hi Tommy, I have explored this, we are already using async validation and not depending on webhook validation. Though webhook exists, but it is noop in my understanding. Please correct me, if this is not the case. 

/cc @Tomcli

**Comment by pugangxa** [comment]
> Hi Tommy, I have explored this, we are already using async validation and not depending on webhook validation. Though webhook exists, but it is noop in my understanding. Please correct me, if this is not the case.
> 
> /cc @Tomcli

Do you mean the code here? https://github.com/kubeflow/kfp-tekton/blob/af9f03d000cc69aaac7aee91361cad3b114347ff/tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go#L183-L188

Yes I think it's good enough now. IMO it's not the same as webhook validation and we can add more logic here if needed in future.

**Comment by Tomcli** [comment]
thanks @ScrapCodes, I think the only concern here is to make sure the `Run` can fail in case the validate method or the custom controller is not running. Which can be handled by the `timeout` feature you are implementing. 

Let me change this issue to add timeout once your PR is merged.

**Comment by ScrapCodes** [comment]
This may also be related: https://github.com/tektoncd/pipeline/issues/3997
